[KEP-4188] update doc for beta (#56238)
kubernetes/website: content/en/docs/reference/node/kubelet-pods-api.md

---
content_type: "reference"
title: Kubelet Pods API
weight: 20
---

{{< feature-state feature_gate_name="PodsAPI" >}}

> The Kubelet Pods API provides a way for node-local components to query information about pods running on the node directly from the kubelet. This increases reliability by removing the dependency on the Kubernetes API server for node-local information and reduces load on the control plane.

Reviewer: ```suggestion
The Kubelet Pods API provides a way for Node-local components to query information about {{< glossary_tooltip text="Pods" term_id="pod" >}} running on the {{< glossary_tooltip term_id="Node" >}} directly from the `kubelet`. This increases reliability by removing the dependency on the Kubernetes API server for node-local information and reduces load on the {{< glossary_tooltip text="control plane" term_id="control-plane" >}}.
```

The API serves the most up-to-date information known locally by the kubelet, derived from its internal cache and reconciliation with the container runtime. It remains available even if the node loses connectivity to the Kubernetes control plane.

If the kubelet's pod sources have not finished their initial synchronization yet (for example, right after the kubelet starts or restarts), every operation returns a gRPC `FAILED_PRECONDITION` error instead of incomplete data. Once synchronization completes, requests are served normally.

## Rate limiting {#rate-limiting}

> `ListPods` and `GetPod` requests are rate limited to protect the kubelet from excessive load. By default, the server allows 100 queries per second with a burst of 10 tokens. Requests that exceed this limit receive a gRPC `RESOURCE_EXHAUSTED` error.

Reviewer: Passive voice: "are rate limited" (passes the "by zombies" test). Consider rewriting with an active subject.

```suggestion
The `kubelet` rate-limits `ListPods` and `GetPod` requests to protect itself from excessive load. By default, the server allows `100` queries per second with a burst of `10` tokens. Requests that exceed this limit receive a gRPC `RESOURCE_EXHAUSTED` error.
```

`WatchPods` streams are not subject to this rate limit; instead, a slow consumer that falls behind on its event stream has its watch connection dropped (see [Metrics](#metrics)).

## Metrics {#metrics}

> The kubelet exposes the following metrics for the Pods API, labelled by `server_api_version` and `status_code`:

Reviewer: Passive voice: "labelled by" (passes the "by zombies" test).

```suggestion
The `kubelet` exposes the following metrics for the Pods API, with labels `server_api_version` and `status_code`:
```

> * `kubelet_pod_requests_total`: cumulative number of requests to the API.

Reviewer: names are matching the latest: https://github.com/kubernetes/kubernetes/pull/140286/changes (not the enhancement docs)

* `kubelet_pod_requests_list_total`: number of requests to the `ListPods` endpoint.
* `kubelet_pod_requests_get_total`: number of requests to the `GetPod` endpoint.
* `kubelet_pod_requests_watch_total`: number of requests to the `WatchPods` endpoint.

> The kubelet also exposes `kubelet_pod_watch_events_dropped_total`, which counts watch events dropped because a client was not consuming its event stream quickly enough.

Reviewer: ```suggestion
The `kubelet` also exposes `kubelet_pod_watch_events_dropped_total`, which counts watch events dropped because a client was not consuming its event stream quickly enough.
```
